internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,433 @@
+	currentTime := metav1.Now()
+	clients := make([]*ConfigurationClientWrapper, 0)
+	for _, clientWrapper := range manager.StaticClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	if !manager.ReplicaDiscoveryEnabled {
+		return clients
+	}
+
+	if currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) &&
+		(manager.DynamicClientWrappers == nil ||
+			currentTime.After(manager.lastFallbackClientRefresh.Time.Add(FallbackClientRefreshExpireInterval))) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		_ = manager.DiscoverFallbackClients(url.Host)
+	}
+
+	for _, clientWrapper := range manager.DynamicClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	return clients
+}
+
+func (manager *ConfigurationClientManager) RefreshClients() {
+	currentTime := metav1.Now()
+	if manager.ReplicaDiscoveryEnabled &&
+		currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		_ = manager.DiscoverFallbackClients(url.Host)
+	}
+}
+
+func (manager *ConfigurationClientManager) DiscoverFallbackClients(host string) error {
+	srvTargetHosts, err := QuerySrvTargetHost(host)
+	if err != nil {
+		klog.Warningf("Fail to build fall back clients %s", err.Error())
+	}
+
+	// Shuffle the list of SRV target hosts
+	for i := range srvTargetHosts {
+		j := rand.Intn(i + 1)
+		srvTargetHosts[i], srvTargetHosts[j] = srvTargetHosts[j], srvTargetHosts[i]
+	}
+
+	newDynamicClients := make([]*ConfigurationClientWrapper, 0)
+	for _, host := range srvTargetHosts {
+		if isValidEndpoint(host, manager.validDomain) {
+			targetEndpoint := "https://" + host
+			client, err := manager.newConfigurationClient(targetEndpoint)
+			if err != nil {
+				return fmt.Errorf("build fallback clients failed")
+			}
+			newDynamicClients = append(newDynamicClients, &ConfigurationClientWrapper{
+				Endpoint:       targetEndpoint,
+				Client:         client,
+				BackOffEndTime: metav1.Time{},
+				FailedAttempts: 0,
+			})
+		}
+	}
+
+	manager.DynamicClientWrappers = newDynamicClients
+	manager.lastFallbackClientRefresh = metav1.Now()
+	return nil
+}
+
+func (manager *ConfigurationClientManager) UpdateClientBackoffStatus(clientWrapper *ConfigurationClientWrapper, successful bool) {
+	if successful {
+		clientWrapper.BackOffEndTime = metav1.Time{}
+		clientWrapper.FailedAttempts = 0
+	} else {
+		clientWrapper.FailedAttempts++
+		clientWrapper.BackOffEndTime = metav1.Time{Time: metav1.Now().Add(calculateBackoffDuration(clientWrapper.FailedAttempts))}
+	}
+}
+
+func QuerySrvTargetHost(host string) ([]string, error) {
+	results := make([]string, 0)
+
+	_, originRecords, err := net.LookupSRV(Origin, TCP, host)
+	if err != nil {
```

Comment:
error handling should same as line 230

Reply:
Do you mean check the error's type?

Reply:
Yes , judge type `net.DNSError` and `IsNotFound`